svsim/README.md

```diff
@@ -0,0 +1,42 @@
+# svsim
+
+`svsim` is a low-level library for compiling and controlling SystemVerilog simulations, currently targeting Verilator and VCS as backends.
+
+## Design
+
+In no particular order, here are some of the design considerations of `svsim`:
+* `svsim` does not know about Chisel. The inputs to `svsim` are SystemVerilog files and the `ModuleInfo` class, which encodes all the information `svsim` needs to generate a test harness for the supplied `SystemVerilog`.
+* `svsim` **is not a testing framework**, and is not opinionated about how a higher-level testing framework should be built. `svsim` should enable SystemVerilog simulation for any testing framework and should not be tied to any particular technology (i.e. Scalatest).
+* `svsim` provides a backend-independent test harness. This means that `svsim`'s generated sources are not impacted by the choice of backend and indeed can be shared between multiple backends. If backend-specific behavior is required, it is provided via a compile-time flag (such as `SVSIM_BACKEND_SUPPORTS_DELAY_IN_PUBLIC_FUNCTIONS`).
+* `svsim` simulations are also backend-independent, meaning that the same code launches and controls a simulation regardless of what backend it was compiled using. 
+* `svsim` backends are declarative. This means a backend only controls which executable is invoked for compilation, and what arguments are provided to that executable (either via command line or environment variables). Backends may also provide arguments to the simulation, but may not have any other specialized logic.
+* Simulations are controlled using an ad-hoc protocol where commands are read from `stdin` and messages are written to `stdout`. This protocol is **not** meant to be an ABI, and can change freely between Chisel verisons.
+* Communication with the simulation is pipelined, so the Scala driver can send multiple commands before processing responses from the simulation. This allows `svsim` to effectively eliminate the overhead of communicating with the simulation without resorting to JNA or Scala-side caching. This optimization can, of course, be defeated if the driver code waits on every command, but we consider this an antipattern.
```

Comment:
I am curious: did you benchmark this? Is it actually as fast as calling into a C-library function directly?

Reply:
Thanks. That makes sense. So I guess most of the SiFive tests are `BasicTester` style where they only wait for a `stop` or a failed `assert`?